### PR TITLE
Add shortcut to quickly add constraints to packages

### PR DESCRIPTION
**Pull request by andreabedini**
This adds a function to the derivations passthru to be able to quickly add constraints to builds.

E.g. if `.#"ghc8107/cardano-ledger-core/0.1.0.0"` is an derivation that aggregates all components of `cardano-ledger-core-0.1.0.0` then

```
.#"ghc8107/cardano-ledger-core/0.1.0.0".passthru.addConstraint "cardano-crypto-class <2.0.0.1, cardano-binary<1.5.0.1"
```
builds the same aggregates with the additional listed constraints.

Usage from the command line is a bit awkward because nix being ... nix, but we can script over it:

```
nix build $(nix eval --raw '.#"ghc8107/cardano-ledger-core/0.1.0.0"' \
    --apply 'd: (d.passthru.addConstraint "cardano-crypto-class <2.0.0.1, cardano-binary<1.5.0.1").drvPath')
```

Unfortunately there's no `nix build --apply` ([yet](https://github.com/NixOS/nix/pull/6333)), so I have to do it in two steps.

### CODE REVIEW

File: nix/builder.nix [hunk 1]
```diff
@@ -51,13 +51,40 @@ let
...
-      constituents = components;
-      # pass through the plan for debugging purposes
-    } // { passthru = { inherit (project) plan-nix; }; }; 
-in build-chap-package
+      components = builtins.map
+        (c: lib.attrsets.recursiveUpdate c {
+          passthru = {
+            addConstraint = constraint:
+              let
+                project' = project.appendModule { cabalProjectLocal = "constraints: ${constraint}"; };
```
Comment (michaelpj) [inline nix/builder.nix:59]: Why not `addCabalConfig` and drop the `constraints: ` bit, so you can just stick whatever additional config line you want in there? Sometimes you might want something else, e.g. an `allow-newer`.

File: nix/builder.nix [hunk 2]
```diff
@@ -51,13 +51,40 @@ let
...
+              in
+              # this is a ugly way to find the same component again
```
Comment (michaelpj) [inline nix/builder.nix:61]: :see_no_evil:

File: nix/builder.nix [hunk 3]
```diff
@@ -51,13 +51,40 @@ let
...
-      # actually run tests: CHaP will not check that your tests pass (neither
-      # does Hackage).
-      constituents = components;
-      # pass through the plan for debugging purposes
-    } // { passthru = { inherit (project) plan-nix; }; }; 
-in build-chap-package
+      components = builtins.map
+        (c: lib.attrsets.recursiveUpdate c {
+          passthru = {
+            addConstraint = constraint:
```
Comment (michaelpj) [inline nix/builder.nix:57]: Should we upstream this, seems kind of useful?

File: nix/builder.nix [hunk 4]
```diff
@@ -51,13 +51,40 @@ let
...
+        (haskellLib.getAllComponents pkg);
+
+      aggregate = pkgs.releaseTools.aggregate
+        {
+          name = package-id;
+          # Note that this does *not* include the derivations from 'checks' which
+          # actually run tests: CHaP will not check that your tests pass (neither
+          # does Hackage).
+          constituents = components;
+          # pass through the plan for debugging purposes
```
Comment (michaelpj) [inline nix/builder.nix:77]: in the wrong place now

**Review by michaelpj — APPROVED** [review]
Needs some doc, at least what you wrote in the PR description should go somewhere :)

File: nix/builder.nix (hunk 2 above)
Comment (andreabedini) [reply nix/builder.nix:61]: Opened https://github.com/input-output-hk/haskell.nix/issues/1951

File: nix/builder.nix (hunk 3 above)
Comment (andreabedini) [reply nix/builder.nix:57]: Yes, eventually :)

**Review by michaelpj — APPROVED** [review]

**Pull request merged**